File: crates/docpie/src/org/schemas.rs
use crate::org::models::Org;
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Serialize, Deserialize)]
pub struct ReadOrg {
    pub name: String,
    pub slug: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ReadResolvedInvitation {
    pub inviter_name: String,
    pub org_name: String,
    pub invite_token: Uuid,
}

impl ReadResolvedInvitation {
    pub fn new(inviter_name: String, org_name: String, invite_token: Uuid) -> Self {
        Self {
            inviter_name,
            org_name,
            invite_token,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Validate)]
pub struct CreateOrg {
    #[validate(length(min = 1, max = 64))]
    pub name: String,

    #[validate(length(min = 3, max = 256))]
    pub slug: String,
}

impl From<Org> for ReadOrg {
    fn from(value: Org) -> Self {
        Self {
            name: value.name,
            slug: value.slug,
        }
    }
}

File: apps/docpie/src/state/org/schemas.ts
import { z } from "zod"
import { constEnum } from "~/lib/const"
import { DisplayName } from "~/state/auth/schemas"

export type TRole = "owner" | "member"

export const Role = constEnum<TRole, { title: string }>({
    owner: {
        value: "owner",
        title: "Owner",
    },
    member: {
        value: "member",
        title: "Member",
    },
})

export const RoleUnion = z.union(
    Object.values(Role).map((role) => z.literal(role.value)),
    { error: (issue) => `"${issue.input}" is not a valid role` },
)

export const OrgName = z
    .string()
    .nonempty("Please enter organization name")
    .max(64, "Exceeds maximum length of 64 characters")
export const OrgSlug = z
    .string()
    .min(3, "Must be at least 3 characters long")
    .max(256, "Exceeds maximum length of 256 characters")

export const Org = z.object({
    name: OrgName,
    slug: OrgSlug,
})

export type TOrgIn = z.input<typeof Org>
export type TOrgOut = z.output<typeof Org>

export const CreateOrg = Org.pick({
    name: true,
    slug: true,
})

export type TCreateOrgIn = z.input<typeof CreateOrg>
export type TCreateOrgOut = z.output<typeof CreateOrg>

export const InviteToken = z.uuid()

export type TInviteTokenIn = z.input<typeof InviteToken>
export type TInviteTokenOut = z.output<typeof InviteToken>

export const NullableInviteToken = InviteToken.nullable()

export type TNullableInviteTokenIn = z.input<typeof NullableInviteToken>
export type TNullableInviteTokenOut = z.output<typeof NullableInviteToken>

export const ResolvedInvitation = z.object({
    org_name: OrgName,
    inviter_name: DisplayName,
    invite_token: InviteToken,
})

export type TResolvedInvitationIn = z.input<typeof ResolvedInvitation>
export type TResolvedInvitationOut = z.output<typeof ResolvedInvitation>
